### ISSUE TITLE
dd.read_hdf with pattern format can be slow

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#1353]
A user reports that read_hdf can be a bit slow when using pattern matching because calling fnmatch many times can be slow.  Perhaps there is a faster way to narrow down key-spaces based on group prefix?

**Comment by alberthdev** [comment]
To add on, I think the major cause of the slowdown is the .keys() call on the internal HDFStore, e.g. https://github.com/dask/dask/blob/d546af029368e1b99e8a45f014185870a8582318/dask/dataframe/io.py#L497. Regardless of the fnmatch, this seems to be the most time-consuming of all. (This might be something to bring up to the PyTables people as well…?)

**Comment by nirizr** [comment]
I tend to agree with @alberthdev, the slowness seems to be related to pytable's `walk_nodes` called by pandas' `HDFStore.groups` method.
Further more it looks like it's an inherent and known drawback with the hdf format, focusing on data rather than metadata and accepting relatively slow metadata acquisition.
There also seems to be some caching involved.

I did see someone recommend storing the needed metadata in it's own hdf node, but we might find other ways to optimize, such as avoiding the 2nd read of each (accepted) node in `get_storer` a couple lines below by calling `hdf.items()` instead of `hdf.keys()`.

**Comment by nirizr** [comment]
The only solution I can think of for this is embedding needed data inside hdf files as a metadata-like dataset.

This will require quite a bit of considerations regarding multiple files. Is the metadata stored in the first file? each file contains it's own? will reading only some of the files be possible?

**Comment by mrocklin** [comment]
Well, we call `hdf.keys()` many times in a single call to read_hdf.  I wonder if perhaps most of the cost here can be avoided by moving this call outside of the `get_keys_and_stops` function within the `_read_single_hdf` function.

**Comment by mrocklin** [comment]
Oh, never-mind, it appears that neither of these functions get called multiple times unnecessarily.

**Comment by nirizr** [comment]
I think I tried something similar, with no improvement:

> such as avoiding the 2nd read of each (accepted) node in get_storer a couple lines below by calling hdf.items() instead of hdf.keys(). I tried using a single iteration on a ~10gig hdf with more than 15k nodes, I experienced no improvement.

More specifically, I wrote a helper function that directly uses pandas' internals to get node names and nrows in one go. It had to improvement, probably due to caching making the second iteration fast.
I can probably dig up the code, if anyone's interested.

**Comment by mrocklin** [comment]
OK, so if a single call to `.keys` is expensive then is there a way to get the necessary group information without calling `.keys`?  This would be doable (I think) with h5py.  Does PyTables offer an internal API that is less expensive?

**Comment by nirizr** [comment]
As I understand it this is a design choice by the HDF group. Reading attributes requires walking the hdf tree, and reading all keys of a group can become a bit heavy. As far as I understand it (and I'd gladly be corrected) there's no easy way around it (except storing the metadata in a separate dataset).

**Comment by alberthdev** [comment]
I think there may potentially be some optimizations when parsing the pattern format. With H5py, I can potentially take a pattern format and do some tweaking to get the keys faster. For instance, if I have something like `BigGroup/SubGroup/*/TinyTable`, with some careful parsing, I can grab the `BigGroup/SubGroup` group, call `.keys()` on it, and then iterate through it to match `TinyTable`.

@nirizr Storing metadata is definitely an option - but I think we should also aim to be able to read "raw" HDF5 files (e.g. non-Pandas created HDF5 files). This would allow us to be able to read other HDF5 files out there, especially those generated by others and by other programs.

On the topic of H5py, I've definitely noticed that using H5py to get the keys tends to be much faster than fetching keys from PyTables. Is PyTables doing something extra that prevents us from getting the tables/groups faster?

**Comment by nirizr** [comment]
I tend to think handling specific cases like `BigGroup/SubGroup/*/TinyTable` is not something we'll want to pursue, but that's just my opinion. 

I did not try removing the pandas/pytables layers while listing the keys. It sounds good but i'm not familiar with h5py enough.

@alberthdev do you think you can work out a small example?

**Comment by mrocklin** [comment]
The case of `group/group//*/table` seems somewhat common to me.  Finding the relevant part of the hierarchy and only iterating through that seems nice.

**Comment by jakirkham** [comment]
In h5py there is the visitor pattern that could be used to do something like this more efficiently. In particular see the [`visit` method of `Group`]( http://docs.h5py.org/en/latest/high/group.html#Group.visit ). Is there something similar that could be done with PyTables or Pandas?
